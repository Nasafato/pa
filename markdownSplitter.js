import { unified } from "npm:unified@10.1.2";
import remarkParse from "https://esm.sh/remark-parse@10.0.2";

export function splitMarkdownIntoChunks(markdown) {
  const ast = unified().use(remarkParse).parse(markdown);
  const chunks = [];
  let currentChunk = [];

  const finishChunk = () => {
    if (currentChunk.length > 0) {
      chunks.push(currentChunk.join("\n"));
      currentChunk = [];
    }
  };

  const visitList = (node, accumulator) => {
    switch (node.type) {
      case "list":
      case "paragraph":
      case "listItem": {
        node.children.forEach((c) => visitList(c, accumulator));
        break;
      }
      case "text": {
        accumulator.push(node.value);
        break;
      }
    }
  };

  const visit = (node) => {
    switch (node.type) {
      case "root": {
        node.children.forEach(visit);
        break;
      }
      case "heading": {
        finishChunk();
        // console.log(node.children);
        const depth = node.depth;
        const heading = Array(depth).fill("#").join("");
        const chunk = [heading, " ", ...node.children.map((c) => c.value)].join(
          ""
        );
        currentChunk.push(chunk);
        break;
      }
      case "link": {
        finishChunk();
        const accumulator = [];
        visitLink(node, accumulator);
        currentChunk.push(accumulator.join(" "));
        break;
      }
      case "code":
        finishChunk();
        currentChunk.push(node.value);
        break;
      case "list": {
        finishChunk();
        const accumulator = [];
        visitList(node, accumulator);
        currentChunk.push(accumulator.map((line) => `- ${line}`).join("\n"));
        break;
      }
      case "paragraph": {
        if (node.children) {
          node.children.forEach(visit);
        } else {
          currentChunk.push(node.value);
        }
        break;
      }
      default: {
        finishChunk();
        if (node.children) {
          node.children.forEach(visit);
        } else if (node.value) {
          currentChunk.push(node.value);
        }
        break;
      }
    }
  };

  const visitLink = (node, accumulator) => {
    switch (node.type) {
      case "link":
        node.children.forEach((c) => visitLink(c, accumulator));
        accumulator.push(node.url);
        break;
      case "text":
        accumulator.push(node.value);
    }
  };

  visit(ast);
  finishChunk();
  return chunks;
}
