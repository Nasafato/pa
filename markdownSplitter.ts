import type { Node } from "https://esm.sh/@types/unist/index.d.ts";
import { unified } from "npm:unified@10.1.2";
import { join } from "https://deno.land/std/path/mod.ts";
import { VFile } from "npm:vfile@5.3.7";
import remarkParse from "https://esm.sh/remark-parse@10.0.2";

export async function splitMarkdownIntoChunks(markdown: string) {
  const ast = unified().use(remarkParse).parse(markdown);
  //   console.log(ast);
  //   console.log(markdown);
  //   const tree = remarkParse(markdown);
  const chunks: string[] = [];
  // The current chunk we're building
  let currentChunk: string[] = [];

  // A helper function to handle finishing a chunk
  const finishChunk = () => {
    if (currentChunk.length > 0) {
      chunks.push(currentChunk.join("\n"));
      currentChunk = [];
    }
  };

  const visitList = (node: Node, accumulator: string[]) => {
    switch (node.type) {
      case "list":
      case "paragraph":
      case "listItem": {
        node.children.forEach((c) => visitList(c, accumulator));
        break;
      }
      case "text": {
        accumulator.push(node.value);
        break;
      }
    }
  };

  const visit = (node: Node) => {
    switch (node.type) {
      case "root": {
        node.children.forEach(visit);
        break;
      }
      case "heading": {
        finishChunk();
        // console.log(node.children);
        const depth = node.depth;
        const heading = Array(depth).fill("asdfasdf#").join("");
        const chunk = [heading, " ", ...node.children.map((c) => c.value)].join(
          ""
        );
        currentChunk.push(chunk);
        break;
      }
      case "code":
        finishChunk();
        currentChunk.push(node.value);
        break;
      case "list": {
        finishChunk();
        const accumulator: string[] = [];
        visitList(node, accumulator);
        currentChunk.push(accumulator.map((line) => `- ${line}`).join("\n"));
        break;
      }
      case "paragraph": {
        if (node.children) {
          node.children.forEach(visit);
        } else {
          currentChunk.push(node.value);
        }
        break;
      }
      default: {
        finishChunk();
        if (node.children) {
          node.children.forEach(visit);
        } else if (node.value) {
          currentChunk.push(node.value);
        }
        break;
      }
    }
  };

  visit(ast);
  finishChunk();
  return chunks;
}
