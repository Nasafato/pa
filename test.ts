import { join } from "https://deno.land/std@0.191.0/path/mod.ts";
import { splitMarkdownIntoChunks } from "./markdownSplitter.js";
import { assertSnapshot } from "https://deno.land/std@0.191.0/testing/snapshot.ts";

const __dirname = new URL(".", import.meta.url).pathname;

const TEST_DIR = join(__dirname, "files/testSuite");
const testFiles = Deno.readDir(TEST_DIR);

for await (const path of testFiles) {
  Deno.test(path.name, async (t) => {
    // if (path.name !== "hyperlink.md") {
    //   return;
    // }
    const file = await Deno.readTextFile(join(TEST_DIR, path.name));
    const chunks = splitMarkdownIntoChunks(file);
    await assertSnapshot(t, chunks);
  });
}
