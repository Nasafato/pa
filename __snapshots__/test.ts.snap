export const snapshot = {};

snapshot[`paragraph.md 1`] = `
[
  "### Tuesday",
  "Kettlebell/conditioning day
" +
    "Morning
" +
    "What's a runner's warmup? Maybe just the Gymnastic bodies warmup
" +
    "2 laps around the park",
]
`;

snapshot[`combination.md 1`] = `
[
  "# Title of the document",
  "Below this are all the subsections.",
  'function messAround() {
    console.log("whoops!");
}',
  "## Subsection 1",
  "- This is the first item of a list
" +
    "- This is a nested list. I wonder if it can be parsed.
" +
    "- This is another item in a list
" +
    "- [ ] This is a todo item
" +
    "- [x] This is a done todo item",
  "### Subsubsection A",
  "And hyperlinks? Does it care about that? ",
  "This is a link to Google https://google.com",
  ".",
  "How does it handle quoted sections? I wonder.",
]
`;

snapshot[`list.md 1`] = `
[
  "Another one",
  "- Item A
- Item B
- Item C
- Item D
- Item E",
]
`;

snapshot[`list2.md 1`] = `
[
  "- my wrists and elbows can't handle the table. Need to do this gradually
" +
    "- reverse bear: pretty hard on the wrists
" +
    "- Supine shoulder extension stretch: definitely can't tough fingertips together, definitely feel it in ulnar side of elbow
" +
    "- Same with supine shoulder extension rocks
" +
    "- Seal stretch: definitely feel it in lumbar spine, posterior pelvic tilt causes stretch in abs",
]
`;

snapshot[`hyperlink.md 1`] = `
[
  "This is a link to Google https://google.com",
]
`;
